File: Python/384.打乱数组.py
from typing import List
import random

class Solution:

    def __init__(self, nums: List[int]):
        self.original = nums
        self.shuffled = self.original.copy()


    def reset(self) -> List[int]:
        self.shuffled = self.original.copy()
        return self.shuffled

    def shuffle(self) -> List[int]:
        for i in range(len(self.shuffled)):
            j = i + random.randrange(len(self.shuffled) - i)
            self.shuffled[i], self.shuffled[j] = self.shuffled[j], self.shuffled[i]
        return self.shuffled



# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()

File: 384.打乱数组.cc
#include <algorithm>
#include <vector>
using std::copy;
using std::swap;
using std::vector;

class Solution {
private:
    vector<int> shuffled;
    vector<int> original;

public:
    Solution(vector<int> &nums) {
        this->shuffled = nums;
        this->original.resize(shuffled.size());
        copy(shuffled.begin(), shuffled.end(), original.begin());
    }

    vector<int> reset() {
        copy(original.begin(), original.end(), shuffled.begin());
        return shuffled;
    }

    vector<int> shuffle() {
        for (int i = 0; i < shuffled.size(); ++i)
        {
            int j = i + rand() % (shuffled.size() - i);
            swap(shuffled[i], shuffled[j]);
        }
        return shuffled;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(nums);
 * vector<int> param_1 = obj->reset();
 * vector<int> param_2 = obj->shuffle();
 */